pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java

```diff
@@ -474,21 +477,31 @@ public byte[] serialize(HyperLogLog hyperLogLog) {
     @Override
     public HyperLogLog deserialize(byte[] bytes) {
       try {
-        return HyperLogLog.Builder.build(bytes);
-      } catch (IOException e) {
-        throw new RuntimeException("Caught exception while de-serializing HyperLogLog", e);
+        IntBuffer intBuf =
+                ByteBuffer.wrap(bytes)
+                        .order(ByteOrder.BIG_ENDIAN)
+                        .asIntBuffer();
+        // The first 2 bytes are constant headers for the HLL. We only need the first byte, the log2m.
+        // We skip the second byte, array size, as we compare the buffer size to the remaining bytes size.
+        int log2m = intBuf.get(0);
+        intBuf.position(2);
+
+        if (intBuf.remaining() != RegisterSet.getSizeForCount(1 << log2m)) {
```

Comment:
Per pure performance concern, this check can be skipped

Reply:
I think we should keep it here, as otherwise we can't guarantee the deserialization is right. For eg, the unit test I added, if you pass a bytes arr that is smaller then expected, it doesnt throw an exception unless this is there. This can also help us catch issue if the underlying HLL library changes that could break this, we'd know